src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs

```diff
@@ -0,0 +1,112 @@
+            _viewBufferScope = viewBufferScope ?? throw new ArgumentNullException(nameof(viewBufferScope));
+        }
+
+        /// <summary>
+        /// The name of the partial view used to create the HTML markup.
+        /// </summary>
+        public string Name { get; set; }
+
+        /// <summary>
+        /// A model to pass into the partial view.
+        /// </summary>
+        public object Model { get; set; }
+
+        /// <summary>
+        /// A <see cref="ViewDataDictionary"/> to pass into the partial view.
+        /// </summary>
+        public ViewDataDictionary ViewData { get; set; }
+
+        [HtmlAttributeNotBound]
+        [ViewContext]
+        public ViewContext ViewContext { get; set; }
+
+        /// <inheritdoc />
+        public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (output == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var viewBuffer = new ViewBuffer(_viewBufferScope, Name, ViewBuffer.PartialViewPageSize);
+            using (var writer = new ViewBufferTextWriter(viewBuffer, Encoding.UTF8))
+            {
+                await RenderPartialViewAsync(writer);
+
+                // Reset the TagName. We don't want `partial` to render.
+                output.TagName = null;
+                output.Content.SetHtmlContent(viewBuffer);
+            }
+        }
+
+        private async Task RenderPartialViewAsync(TextWriter writer)
+        {
+            var viewEngineResult = _viewEngine.GetView(ViewContext.ExecutingFilePath, Name, isMainPage: false);
+            var getViewLocations = viewEngineResult.SearchedLocations;
+            if (!viewEngineResult.Success)
+            {
+                viewEngineResult = _viewEngine.FindView(ViewContext, Name, isMainPage: false);
+            }
+
+            if (!viewEngineResult.Success)
+            {
+                var searchedLocations = Enumerable.Concat(getViewLocations, viewEngineResult.SearchedLocations);
+                var locations = string.Empty;
+                if (searchedLocations.Any())
+                {
+                    locations += Environment.NewLine + string.Join(Environment.NewLine, searchedLocations);
+                }
+
+                throw new InvalidOperationException(
+                    Resources.FormatViewEngine_PartialViewNotFound(Name, locations));
+            }
+
+            var view = viewEngineResult.View;
+            using (view as IDisposable)
+            {
+                // Determine which ViewData we should use to construct a new ViewData
+                var baseViewData = ViewData ?? ViewContext.ViewData;
+
+                var newViewData = new ViewDataDictionary<object>(baseViewData, Model);
```

Comment:
Overloads of the `PartialAsync(...)` and `RenderPartialAsync(...)` helpers default the model they use to `htmlHelper.ViewData.Model`. This instead defaults to `null`, which is unexpected. Should do something similar to the HTML helpers here.

Always use `ViewContext.ViewData.Model` since setting the `ViewData` property should not change the `Model` property's default.